src/charm.py

```diff
@@ -2,203 +2,144 @@
+            "summary": "spark layer",
+            "description": "pebble config layer for spark-k8s",
+            "services": {
+                self._container_name: {
+                    "override": "replace",
+                    "summary": "Spark Operator layer",
+                    "startup": "enabled",
+                    "command": (
+                        f"/usr/bin/tini -s -- /usr/bin/spark-operator -v=2 "
+                        "-logtostderr "
+                        f"-namespace={self.model.name} "
+                        "-enable-ui-service=true "
+                        "-controller-threads=10 "
+                        "-resync-interval=30 "
+                        "-enable-batch-scheduler=false "
+                        "-enable-metrics=true "
+                        "-metrics-labels=app_type "
+                        f"-metrics-port={self.model.config['metrics-port']} "
+                        "-metrics-endpoint=/metrics "
+                        "-enable-resource-quota-enforcement=false "
+                        "-enable-webhook=true "
+                        f"-webhook-svc-namespace={self.model.name} "
+                        f"-webhook-port={self.model.config['webhook-port']} "
+                        f"-webhook-svc-name={self.model.app.name} "
+                        f"-webhook-config-name={self._mutating_webhook_name} "
+                        f"-webhook-namespace-selector=model.juju.is/name={self.model.name} "
+                        "-webhook-fail-on-error=true"
+                    ),
                 }
             },
+        }
+        return pebble_layer
+
+    def _on_spark_pebble_ready(self, _):
+        self.unit.status = MaintenanceStatus("Configuring Spark Charm")
+
+        self.container.push("/etc/webhook-certs/ca-cert.pem", self._stored.ca, make_dirs=True)
+        self.container.push(
+            "/etc/webhook-certs/server-cert.pem", self._stored.cert, make_dirs=True
         )
-        self.model.unit.status = ActiveStatus()
+        self.container.push("/etc/webhook-certs/server-key.pem", self._stored.key, make_dirs=True)
+
+        self.resource_handler.apply()
+
+        self.container.add_layer(self._container_name, self._pebble_layer, combine=True)
+        self.container.autostart()
+
+        self.unit.status = ActiveStatus()
+
+    def _on_config_changed(self, event):
+        # check connection; config change could come before pebble ready
+        if not self.container.can_connect():
+            self.unit.status = WaitingStatus("Waiting to connect to spark container")
+            event.defer()
+            return
+
+        self.unit.status = MaintenanceStatus("Updating Spark Charm's Configurations")
+        self.service_patcher._patch(self.service_patcher)
+
+        self.container.add_layer(self._container_name, self._pebble_layer, combine=True)
+        self.container.replan()
+
+        self.unit.status = ActiveStatus()
+
+    def _on_remove(self, _):
```

Comment:
@DomFleischmann is this another charm where juju is supposed to remove all the resources?

Reply:
It technicaly is, but we also rewrote this charm because the charm was leaving behind some resources whch would end up blocking any future deployments. I think that is why @agathanatasha decided on removing everything.

Reply:
Juju is definitely not removing the mutating webhook which causes problems. Previously I didn't named the clusterrole with the model name and that also caused errors on install. I thought it would be nice to remove everything for tidiness sake

Reply:
@DomFleischmann I don't understand when juju is supposed to do all the work and when it's fine to "manually" remove all the resources in the remove event handler. I've raised this concern in the past and it is still not clear to me. We have other charms that create their own k8s resources, and when removed, they leave all resources in the cluster. What considerations should we take in when writing these type of charms?

Reply:
@DnPlas the confusion is understandable, so far we have been avoiding removing things in the remove hook due to the juju team telling us it is a bug. At this point, I am concerned that this will not be fixed in time until we deploy sidecar charms on a larger scale. So we must start removing things on our own until the juju team fixes these issues.